Add code snippet for Generate random numbers

asilenko opened the pull request:
Added code snippet for Issue #105 :
- created two methods: 
   -> for random number between x and y
   -> for throwing dice
- added tests for both methods,
- updated Readme.md

iluwatar commented on src/main/java/math/GenerateRandomNumbersSnippet.java:
```diff
@@ -0,0 +1,56 @@
+     * @param y right bound
+     * @return random integer between x and y
+     */
+    public static int generateRandomNumberBetweenXAndY(int x, int y) {
```
I would suggest `generateRandomNumberBetween` (leave out the end)

asilenko replied:
Done

iluwatar commented on src/main/java/math/GenerateRandomNumbersSnippet.java:
```diff
@@ -0,0 +1,56 @@
+        return random.nextInt(exclusiveRightBound) + x;
+    }
+
+    /**
+     * Throw dice (3d6)
+     *
+     * @return random integer between 1 and 6 (inclusive)
+     */
+    public static int throwDice() {
+        return random.nextInt(5) + 1;
+    }
```
Maybe the function could take parameters of how many dice we have and what kind of dice it is. For example, parameters 3 and 6 would throw 3 times d6 dice giving a result between 3-18.

asilenko replied:
Done

iluwatar commented on src/test/java/math/GenerateRandomNumbersSnippetTest.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Tests for {@link GenerateRandomNumbersSnippet#generateRandomNumberBetweenXAndY(int, int)}.
+     */
+    @Test
+    void testGenerateRandomNumberBetweenXAndY() {
+        var randomInteger0 = GenerateRandomNumbersSnippet.generateRandomNumberBetweenXAndY(0, 0);
+        assertEquals(0, randomInteger0);
+
+        var randomInteger1 = GenerateRandomNumbersSnippet.generateRandomNumberBetweenXAndY(3, 10);
+        var listOfPossibleResults1 = List.of(3, 4, 5, 6, 7, 8, 9, 10);
+        assertTrue(listOfPossibleResults1.contains(randomInteger1));
+
+        var randomInteger2 = GenerateRandomNumbersSnippet.generateRandomNumberBetweenXAndY(42, 43);
+        var listOfPossibleResults2 = List.of(42, 43);
+        assertTrue(listOfPossibleResults2.contains(randomInteger2));
+    }
+
+    /**
+     * Tests for {@link GenerateRandomNumbersSnippet#throwDice()}.
+     */
+    @Test
+    void testThrowDice() {
+        var number = GenerateRandomNumbersSnippet.throwDice();
+        var listOfPossibleResults = List.of(1, 2, 3, 4, 5, 6);
+        assertTrue(listOfPossibleResults.contains(number));
+    }
```
I think parameterized tests would be very useful here to test more thoroughly. See https://www.baeldung.com/parameterized-tests-junit-5

asilenko replied:
Parameterized tests added.  I'm new to junit, thanks for the opportunity to learn.

iluwatar reviewed (changes requested):
In addition to the review comments, there is a merge conflict that needs to be resolved.

asilenko commented:
Ready for review.

iluwatar reviewed (changes requested):
You need to fix the failing build

asilenko commented:
> You need to fix the failing build

I had some trouble working with checkstyle.xml It seems to be OK by now. 
In addition I added one new test to check if all numbers from given range were generated.

iluwatar reviewed (changes requested):
The build is still failing due to Checkstyle issue. Please set up your IDE to work with Checkstyle and run the checks locally before submitting changes for review.

asilenko commented:
Sorry for the delay.  Hope this last commit has fixed the problem with the checkstyle. I'm new to gradle. The project still do not build on my device but the reason is build.gradle file. I'm not sure if I should fix it along with this issue. Let me know if it's ok.

iluwatar reviewed (changes requested):
The build is still failing due to the missing license header and Checkstyle issues.

iluwatar commented:
@asilenko I added some more instructions in the PR https://github.com/iluwatar/30-seconds-of-java/pull/119 Maybe you find that helpful.

asilenko commented:
> @asilenko I added some more instructions in the PR #119 Maybe you find that helpful.

Thank you. I fixed the licence, but I don't see any checkstyle warning when I try to build the project. If it still does not work just pass the task to someone else. Sorry for wasting your time.

grzesiekkedzior commented:
Ok, @asilenko I checked your code very fast and in my Idea everything compile successfully... almost. In class GenerateRandomNumbersSnippetTest change line 67. Line is longer than 100 characters. In checkstyle.xml at line 34 the max line length value equals 100. Line 67 has 102.

iluwatar commented on README.md:
````diff
@@ -480,8 +481,9 @@ Inspired by [30 seconds of code](https://github.com/Chalarangelo/30-seconds-of-c
+
 ```java
-public static int calculateLuhnChecksum(long num) {
+  public static int calculateLuhnChecksum(long num) {
````
No need to modify this line

iluwatar commented on README.md:
```diff
@@ -504,6 +506,25 @@ public static int calculateLuhnChecksum(long num) {
+  private static final Random random = new Random();
+
+  public static int generateRandomNumberBetween(int x, int y) {
+    int exclusiveRightBound = y - x + 1;
```
use `var` keyword where possible

iluwatar commented on src/test/java/math/GenerateRandomNumbersSnippetTest.java:
```diff
@@ -0,0 +1,95 @@
+    assertTrue(randomInteger >= min && randomInteger <= max);
+  }
+
+  @Test
+  void checkThatAllNumbersInTheSpecifiedRangeHaveBeenGenerated() {
+    List<Integer> listOfResults = new ArrayList<Integer>();
+    for (int i = 0; i < 200; i++) {
+      listOfResults.add(GenerateRandomNumbersSnippet.generateRandomNumberBetween(1, 10));
+    }
+    List<Integer> listOfDistinctResults =
+        listOfResults.stream()
+            .distinct()
+            .sorted(Integer::compareTo)
+            .collect(Collectors.toList());
+
+    List<Integer> expectedResult = List.of(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+
+    assertEquals(listOfDistinctResults, expectedResult);
+  }
```
This test looks like it can randomly fail. I'm not sure I understand the purpose of the test either.

iluwatar commented on src/main/java/math/GenerateRandomNumbersSnippet.java:
```diff
@@ -0,0 +1,62 @@
+   *
+   * @param numberOfDice represents number of dice being thrown
+   * @param typeOfDice   represents number of sides of the dice
+   * @return random integer between 1 and 6 (inclusive)
```
Not true, depends on the parameters

iluwatar commented on src/test/java/math/GenerateRandomNumbersSnippetTest.java:
```diff
@@ -0,0 +1,95 @@
+  /**
+   * Tests for {@link GenerateRandomNumbersSnippet#throwDice(int, int)}.
+   */
+  @CsvSource({
+      "1, 6",
+      "2, 6",
+      "3, 6",
+      "1, 10",
+      "2, 10",
+      "10, 6",
+      "10, 10",
+      "20, 5"
+  })
```
The test data is lacking 0 and negative numbers

PraveenNanda124 reviewed (commented):
Good work
